### PR TITLE
ASoC: Intel: sof_sdw: Add dummy codec for BT audio offload

### PR DESCRIPTION
**Pull request by yongzhi1**
Signed-off-by: Yong Zhi <yong.zhi@intel.com>

### CODE REVIEW

**Comment by yongzhi1** [comment]
This is the same feature as of PR #2773 , code has been tested with topology changes, looking for feedback, thanks!!

File: sound/soc/intel/common/soc-acpi-intel-tgl-match.c [hunk 1]
```diff
@@ -394,6 +394,7 @@ struct snd_soc_acpi_mach snd_soc_acpi_intel_tgl_sdw_machines[] = {
 	},
 	{
 		.link_mask = 0x3, /* rt5682 on link0 & 2xmax98373 on link 1 */
+		.id = "CnviBt",
```
Comment (plbossart) [inline sound/soc/intel/common/soc-acpi-intel-tgl-match.c]: what is this? an ACPI _HID? If yes, that's not a compliant HID and thus a NAK from me.

File: sound/soc/intel/boards/sof_sdw.c [hunk 2]
```diff
@@ -379,6 +379,11 @@ static struct sof_sdw_codec_info codec_info_list[] = {
 		.dai_name = "rt5682-sdw",
 		.init = sof_sdw_rt5682_init,
 	},
+	{
+		.acpi_id = "CnviBt",
```
Comment (plbossart) [inline sound/soc/intel/boards/sof_sdw.c]: you're killing me.

File: sound/soc/intel/boards/sof_sdw.c [hunk 3]
```diff
@@ -1066,9 +1074,11 @@ static int sof_card_dai_links_create(struct device *dev,
...
 			      NULL, info->ops);
 
-		ret = info->init(NULL, links + link_id, info, 0);
-		if (ret < 0)
-			return ret;
+		if (info->init) {
+			ret = info->init(NULL, links + link_id, info, 0);
+			if (ret < 0)
+				return ret;
+		}
```
Comment (plbossart) [inline sound/soc/intel/boards/sof_sdw.c]: should be a separate patch.

File: sound/soc/intel/boards/sof_sdw.c [hunk 4]
```diff
@@ -146,8 +146,8 @@ static const struct dmi_system_id sof_sdw_quirk_table[] = {
 			DMI_MATCH(DMI_PRODUCT_NAME, "Volteer"),
 		},
 		.driver_data = (void *)(SOF_SDW_TGL_HDMI |
-					SOF_SDW_PCH_DMIC |
-					SOF_SDW_FOUR_SPK),
+					SOF_SDW_PCH_DMIC | SOF_SDW_FOUR_SPK |
+					SOF_SSP_PORT(SOF_I2S_SSP2)),
```
Comment (plbossart) [inline sound/soc/intel/boards/sof_sdw.c]: that probably works but that macro was intended for amplifiers.
We should probably have a different macro to support the case of SDW headset + I2S amplifier + BT offload. not sure if it makes sense but it's better to plan ahead.

**Review by plbossart — CHANGES_REQUESTED** [review]

File: sound/soc/intel/boards/sof_sdw.c (hunk 4 above)
Comment (yongzhi1) [reply sound/soc/intel/boards/sof_sdw.c]: Ack, a new macro to distinguish I2S used for BT-offload purpose is the way to go. 
With the new macro, we also have the advantage to maintain the link id order in topology instead of inserting BT-offload before DMIC.

File: sound/soc/intel/boards/sof_sdw.c (hunk 2 above)
Comment (yongzhi1) [reply sound/soc/intel/boards/sof_sdw.c]: I know, I know, just to make sure this is no-go.

(Although made obvious, I probably should explain upfront this is a fake id to be used by find_codec_info_acpi() only)

File: sound/soc/intel/boards/sof_sdw.c (hunk 3 above)
Comment (yongzhi1) [reply sound/soc/intel/boards/sof_sdw.c]: Ack.

File: sound/soc/intel/common/soc-acpi-intel-tgl-match.c (hunk 1 above)
Comment (yongzhi1) [reply sound/soc/intel/common/soc-acpi-intel-tgl-match.c]: Just use the ACPI_HID field to illustrate how the code works in the current patch - there is no real "CnviBt" programmed at coreboot/bios.

In other words, a new macro can be used to avoid using find_codec_info_acpi() for the BT offload dai link.

File: sound/soc/intel/boards/sof_sdw.c [hunk 5]
```diff
@@ -1054,6 +1059,9 @@ static int sof_card_dai_links_create(struct device *dev,
 		if (!codec_name)
 			return -ENOMEM;
 
+		if (!strcmp(info->dai_name, "snd-soc-dummy-dai"))
+			codec_name = "snd-soc-dummy";
```
Comment (plbossart) [inline sound/soc/intel/boards/sof_sdw.c]: move at line 1057 and only allocate the codec_name in an else case then. No need to allocate and not use

File: sound/soc/intel/common/soc-acpi-intel-tgl-match.c (hunk 1 above)
Comment (plbossart) [reply sound/soc/intel/common/soc-acpi-intel-tgl-match.c]: Goodness. 
I have mixed feelings about this. On one hand, it's clearly an abuse of what this id was intended for, but on the other hand this hack is quite brilliant. I would never have thought about this...
Need to think about this, I am still not sold on the idea that we want to unconditionally add BT offload, we'd probably need something from platform firmware telling us to enable this link.

Comment (andy-shev) [reply sound/soc/intel/common/soc-acpi-intel-tgl-match.c]: I am not opposing the "hack" part, but what it must do is to **make sure** that user won't have an impression that this is a _HID() and, if it suddenly appears in firmware or bootloader generated ACPI table, this can possible work. I.o.w. if you guarantee that
- user won't mistakenly recognize this as ACPI _HID()
- this won't work in case the firmware / bootloader ACPI table has it

But I see a problem here. If by any chance somebody will abuse ACPI so much and actually put this as a name in the table, you will be in a position of having a collision of the names.

That said, it mustn't be anyhow in the fields that may match with real _HID():s.

Comment (plbossart) [reply sound/soc/intel/common/soc-acpi-intel-tgl-match.c]: I also wonder if this is a platform-level information that needs to be stored in a table, or something that can be discovered based on CNVI PCI ID and/or flags/capabilities.

Comment (yongzhi1) [reply sound/soc/intel/common/soc-acpi-intel-tgl-match.c]: So here are the 3 options:

1. Use this field with the fake non ACPI compliant name intentionally + add new quirk macro
2. Use quirk macro alone to resolve the DAI link as done in https://github.com/thesofproject/linux/pull/2774/commits/50b4529b2974d03ae75c1c2ab4451b4dde443ed0
3. Use genuine existing ACPI attribute if available, or add new one in cb for query.

Here is the BT-offload support implemented in coreboot:
https://review.coreboot.org/c/coreboot/+/45586/4/src/mainboard/google/volteer/variants/baseboard/devicetree.cb

Since machine driver is not board-blind, I feel option 1 and 2 are sufficient, but will follow the consensus reached here.

**Comment by kv2019i** [comment]
@plbossart I checked the NHLT tables in some systems supporting BT audio offload on Windows and there's no really info about the BT offload. Rather it just has a SSP configuration, just like any other SSP codec. So based on NHLT, we can't tell whether BT-audio-offload is supposed to be used or not.

So for now, I suggest we treat this as a generic SSP codec. I.e. a specific machine driver is needed that knows about the configuration (and there's a matching tplg file). The latest revision from Yong should now be aligned with this.

File: sound/soc/intel/boards/sof_sdw.c (hunk 3 above)
Comment (kv2019i) [reply sound/soc/intel/boards/sof_sdw.c]: This is handled in latest revision, marking as resolved.

File: sound/soc/intel/boards/sof_sdw.c (hunk 2 above)
Comment (kv2019i) [reply sound/soc/intel/boards/sof_sdw.c]: This is handled in latest revision, marking as resolved.

File: sound/soc/intel/boards/sof_sdw.c [hunk 6]
```diff
@@ -1136,6 +1140,20 @@ static int sof_card_dai_links_create(struct device *dev,
...
+						GFP_KERNEL);
+		if (!ssp_components)
+			return -ENOMEM;
+		/* BT to SSP is fixed on Port 2 for TGL/ADL platforms */
+		ssp_components->name = "snd-soc-dummy";
+		ssp_components->dai_name = "snd-soc-dummy-dai";
+		cpus[cpu_id].dai_name = "SSP2 Pin";
+
+		init_dai_link(dev, links + link_id, be_id, "SSP2-NoCodec", 1, 1,
+				cpus + cpu_id, 1, ssp_components, 1, NULL, NULL);
```
Comment (plbossart) [inline sound/soc/intel/boards/sof_sdw.c:1192]: @yongzhi1 do you mind updating the code to use the same convention as with tgl-max98373 (SSP2-BT) and quirks? Thanks.

Comment (yongzhi1) [reply sound/soc/intel/boards/sof_sdw.c:1192]: Sure, sorry about the delayed response.

**Comment by yongzhi1** [comment]
v3 update:

1. Use quirk macros to represent the SSP port assignment for BT offload
2. Rename BE codec name from "SSP#-NoCodec" to "SSP#-BT".

Tested with https://github.com/yongzhi1/sof/commit/0defd74b4bc34d9464dee281027013ea4ed3dc85

**Review by kv2019i — APPROVED** [review]
Thanks @yongzhi1 !

**Review by plbossart — APPROVED** [review]
